File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    /*
    ["((( )))","(( )( ))","(( ))()","()(( ))","()()()"]
    */
   let ans = [];
    function dfs(str, open, close){
        if(open < close) return;
        
        if(open === n && close === n){
            ans.push(str);
            return;
        }
        
        if(open<n) dfs(str+'(', open+1, close);
        if(close < n) dfs(str+')', open, close+1);
    }
    
    dfs('', 0 ,0);
    return ans;
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        '''
        Input: n = 3
        Output: ["((()))","(()())","(())()","()(())","()()()"]
        
        only add open brackets if open < n
        only add a closing bracked if closed < open
        valid if open == closed == n
        
        '''
        stack = []
        ans = []
        
        def backtrack(openN, closedN):
            if openN == closedN == n:
                ans.append("".join(stack))
                return
            if openN < n:
                stack.append("(")
                backtrack(openN + 1, closedN)
                stack.pop()
                
            if closedN < openN:
                stack.append(")")
                backtrack(openN, closedN+1)
                stack.pop()
                
        backtrack(0,0)
        return ans